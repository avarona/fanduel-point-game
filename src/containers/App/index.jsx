import React from 'react';
import { fetchPlayers } from '../../api';
import { selectRandom } from '../../utils';
import { AppContextProvider } from '../../AppContext';

import styles from './styles.module.css';
import Match from '../../components/Match';

class App extends React.Component {
  state = {
    loading: true, // app loading state
    players: [], // player pool from api call
    matchPlayers: [], // "tuple" of 2 players head to head
    winnerId: undefined, // id of player with higher fppg
    correctGuessCount: 0 // user's correct guess count
  }
  
  componentDidMount() {    
    // fetch and store players to App's state
    fetchPlayers.then(({ data }) => {
      const { players } = data;
      this.setState({ players }, () => {
        this.createMatch(); // create the match once players are retrieved from backend
        this.setState({ loading: false }); // set loading state after the players are set to the app state
      });
    });
  }
  
  /**
   * Creates a new match of 2 random players
   */
  createMatch = () => {
    const { players } = this.state;
    const matchPlayers = selectRandom(players, 2);
    this.setState({ matchPlayers, winnerId: undefined });
  }
  
  /**
   * Click handler for selecting the user's guess
   * 
   * @param id - player id
   */
  selectPlayer = (id) => {
    const { matchPlayers, correctGuessCount } = this.state;
    const winnerId = this.calculateWinner(...matchPlayers);

    if(winnerId === id) this.setState({ correctGuessCount: correctGuessCount + 1 });
  };
  
  /**
   * Determines the player with the highest fppg and sets him as the winner
   * 
   * @param player1 - player 1 obj
   * @param player2 - player 2 obj
   */
  calculateWinner = (player1, player2) => {
    if(player1.fppg === player2.fppg) this.setState({ winnerId: -1 });

    const winnerId = player1.fppg > player2.fppg ? player1.id : player2.id

    this.setState({ winnerId });

    return winnerId;
  }

  /**
   * Resets guess count & creates a new match
   */
  resetGame = () => this.setState({ correctGuessCount: 0 }, this.createMatch);

  /**
   * Sets the context value to be fed to the Provider below
   */
  get contextValue() {
    return { 
      ...this.state,
      selectPlayer: this.selectPlayer,
      nextMatch: this.createMatch,
      resetGame: this.resetGame
    }
  }

  render() {
    const { loading, matchPlayers } = this.state;

    return (
      <AppContextProvider value={this.contextValue}>
        <h1 className={styles.heading}>Who has the greater FPPG</h1>
        <main className={styles.appContainer}>
          {loading 
            ? <div>Loading...</div>
            : <Match players={matchPlayers} />}
        </main>
      </AppContextProvider>
    )
  }
}

export default App;
