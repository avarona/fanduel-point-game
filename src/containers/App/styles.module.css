.appContainer {
  display: flex;
  flex-direction: column;
  align-items: center;
  margin-top: 48px;
}

.heading {
  text-align: center;
}
