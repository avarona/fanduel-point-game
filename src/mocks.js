export const players = [
    {id: 1, first_name: 'Kobe', last_name: 'Bryant', fppg: 15, images: { default: { url: 'slick pic' }}},
    {id: 2, first_name: 'Lebron', last_name: 'James', fppg: 10, images: { default: { url: 'the dunker' }}},
    {id: 3, first_name: 'Michael', last_name: 'Jordan', fppg: 5, images: { default: { url: 'the legend' } }}
];

export const matchPlayers = [players[0], players[1]];

export const playerDetails = players[0];