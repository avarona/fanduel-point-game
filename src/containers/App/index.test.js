import React from 'react';
import { shallow } from 'enzyme';
import App from './';
import * as Mocks from '../../mocks';

describe('App Container', () => {
  it('renders', () => {
    const wrapper = shallow(<App />);
    expect(wrapper.html()).toMatchSnapshot();
  });

  describe('methods', () => {
    const wrapper = shallow(<App />);
    const instance = wrapper.instance();
    wrapper.setState({ players: Mocks.players });

    afterEach(() => {
      wrapper.setState({ matchPlayers: [], winnerId: undefined, correctGuessCount: 0 });
    });

    describe('cleanPlayerData(players)', () => {
      it('returns a new players array without null values', () => {
        instance.cleanPlayerData(Mocks.uncleanPlayers);
        expect(instance.cleanPlayerData(Mocks.uncleanPlayers)).toEqual(Mocks.players);
      });
    });

    describe('createMatch()', () => {
      it('sets `matchPlayers` to an array of 2', () => {
        instance.createMatch();
        const matchedPlayers = wrapper.state('matchPlayers');

        expect(matchedPlayers.length).toEqual(2);
        expect(Mocks.players.some(p => matchedPlayers[0].id)).toBe(true);
        expect(Mocks.players.some(p => matchedPlayers[1].id)).toBe(true);
      });
    });

    describe('selectPlayer(id)', () => {
      beforeEach(() => {
        instance.createMatch();
      })

      it('correctGuessCount starts at 0', () => {
        expect(wrapper.state('correctGuessCount')).toEqual(0);
      });
      
      
      it('adds a correct guess', () => {
        const mp = wrapper.state('matchPlayers');
        const winner = mp[0].fppg > mp[1].fppg ? mp[0] : mp[1];
        instance.selectPlayer(winner.id);
        
        expect(wrapper.state('correctGuessCount')).toEqual(1);
      });

      it('does not add a wrong guess', () => {
        const mp = wrapper.state('matchPlayers');
        const loser = mp[0].fppg < mp[1].fppg ? mp[0] : mp[1];
        instance.selectPlayer(loser.id);

        expect(wrapper.state('correctGuessCount')).toEqual(0);
      });
    });

    describe('calculateWinner(player1, player2)', () => {
      it('returns the id of the player with a higher fppg', () => {
        const [player1, player2] = Mocks.players;
        instance.calculateWinner(player1, player2);

        expect(wrapper.state('winnerId')).toEqual(player1.id);
      });
    });

    describe('resetGame()', () => {
      beforeEach(() => {
        instance.createMatch();
      });
      
      it('sets the correctGuessCount to 0', () => {
        instance.resetGame();
        expect(wrapper.state('correctGuessCount')).toEqual(0);
      });

      it('creates a new match', () => {
        instance.resetGame();
        expect(wrapper.state('matchPlayers').length).toBe(2);
      });
    });
  });
});
