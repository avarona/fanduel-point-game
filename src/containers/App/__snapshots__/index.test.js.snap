// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Container renders 1`] = `"<h1 class=\\"heading\\">Who has the greater FPPG?</h1><main class=\\"appContainer\\"><div>Loading...</div></main>"`;
